src/System.Linq/src/System/Linq/Select.SpeedOpt.cs

```diff
@@ -157,6 +157,127 @@ public TResult TryGetLast(out bool found)
+                _selector = selector;
+            }
+
+            public override Iterator<TResult> Clone() =>
+                new SelectRangeIterator<TResult>(_start, _end, _selector);
+
+            public override bool MoveNext()
+            {
+                if (_state < 1 || _state == (_end - _start + 1))
+                {
+                    Dispose();
+                    return false;
+                }
+
+                int index = _state++ - 1;
+                _current = _selector(_start + index);
+                return true;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector) =>
+                new SelectRangeIterator<TResult2>(_start, _end, CombineSelectors(_selector, selector));
+
+            public TResult[] ToArray()
+            {
+                var results = new TResult[_end - _start];
+                int srcIndex = _start;
+                for (int i = 0; i < results.Length; i++)
+                {
+                    results[i] = _selector(srcIndex++);
+                }
+
+                return results;
+            }
+
+            public List<TResult> ToList()
+            {
+                var results = new List<TResult>(_end - _start);
+                for (int i = _start; i != _end; i++)
+                {
+                    results.Add(_selector(i));
+                }
+
+                return results;
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                // In case someone uses Count() to force evaluation of the selector,
+                // run it provided `onlyIfCheap` is false.
+                if (!onlyIfCheap)
+                {
+                    for (int i = _start; i != _end; i++)
+                    {
+                        _selector(i);
+                    }
+                }
+
+                return _end - _start;
+            }
+
+            public IPartition<TResult> Skip(int count)
+            {
+                if (count >= (_end - _start))
+                {
+                    return EmptyPartition<TResult>.Instance;
+                }
+
+                return new SelectRangeIterator<TResult>(_start + count, _end, _selector);
+            }
+
+            public IPartition<TResult> Take(int count)
+            {
+                if (count >= (_end - _start))
+                {
+                    return this;
+                }
+
+                return new SelectRangeIterator<TResult>(_start, _start + count, _selector);
+            }
+
+            public TResult TryGetElementAt(int index, out bool found)
+            {
+                if ((uint)index < (uint)(_end - _start))
+                {
+                    found = true;
+                    return _selector(_start + index);
+                }
+
+                found = false;
+                return default;
+            }
+
+            public TResult TryGetFirst(out bool found)
```

Comment:
I understand that you are not introducing a new pattern here, but using an existing one instead.

Why is this pattern `T Try(out bool)` instead of `bool Try(out T)` ? I wonder if there is any perf-related reason for it.

Reply:
Returning the T rather than having it be an out avoids a GC write barrier if T is or contains a reference type.

Reply:
@stephentoub thanks for the explanation!